include/astshim/FrameDict.h

```diff
@@ -0,0 +1,310 @@
+    */
+    std::shared_ptr<Mapping> getMapping(int from, std::string const &to) const {
+        return FrameSet::getMapping(from, getIndex(to));
+    }
+    std::shared_ptr<Mapping> getMapping(std::string const &from, int to) const {
+        return FrameSet::getMapping(getIndex(from), to);
+    }
+    std::shared_ptr<Mapping> getMapping(std::string const &from, std::string const &to) const {
+        return FrameSet::getMapping(getIndex(from), getIndex(to));
+    }
+    //@}
+
+    /**
+    Get the index of a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    int getIndex(std::string const &domain) const {
+        std::string domainUppercase(domain.size(), ' ');
+        std::transform(domain.begin(), domain.end(), domainUppercase.begin(),
+                   [](unsigned char c) -> unsigned char { return std::toupper(c); });
+        auto it = _domainIndexDict.find(domainUppercase);
+        if (it == _domainIndexDict.end()) {
+            throw std::range_error("No frame found with domain " + domain);
+        }
+        return it->second;
+    }
+
+    /**
+    Return True if a frame in this FrameDict has the specified domain
+    */
+    bool hasDomain(std::string const &domain) const { return bool(_domainIndexDict.count(domain) > 0); }
+
+    using FrameSet::mirrorVariants;
+
+    /**
+    Call mirrorVariants with a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void mirrorVariants(std::string const &domain) { FrameSet::mirrorVariants(getIndex(domain)); }
+
+    using FrameSet::remapFrame;
+
+    /**
+    Remap a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void remapFrame(std::string const &domain, Mapping &map) { FrameSet::remapFrame(getIndex(domain), map); }
+
+    using FrameSet::removeFrame;
+
+    /**
+    Remove a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void removeFrame(std::string const &domain) {
+        FrameSet::removeFrame(getIndex(domain));
+        _rebuildDict();
+    }
+
+    using FrameSet::setBase;
+
+    /**
+    Set the base frame to a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void setBase(std::string const &domain) { FrameSet::setBase(getIndex(domain)); }
+
+    using FrameSet::setCurrent;
+
+    /**
+    Set the current frame to a frame specified by domain
+
+    @throw std::range_error if no frame found with the specified domain
+    */
+    void setCurrent(std::string const &domain) { FrameSet::setCurrent(getIndex(domain)); }
+
+    /// Set the domain of the current frame (and update the internal dict)
+    void setDomain(std::string const &domain) override {
+        Frame::setDomain(domain);
+        _rebuildDict();
+    }
```

Comment:
I tried that and decided it wasn't worth the fuss. It's quite easy to determine in advance if rebuilding is safe, so I do that. I then assert if the rebuild fails since this indicates a logic error and an inconsistent state.

As to why I am rebuilding the entire dictionary: I consider it simpler and safer to have just one or two ways to manipulate the dict: a full rebuild or add a frame.